### ISSUE TITLE
Hill # NaN in genetic and trait results

### ISSUE THREAD

**Issue by isaacovercast** [role-model/roleR#121]
Run this:
```library(roleR)
## init_type may be bridge_island or oceanic_island
p <- roleParams(individuals_local = 100, individuals_meta = 10000,
                species_meta = 10, speciation_local = 0, 
                speciation_meta = 0.05, extinction_meta = 0.05, env_sigma = 0.5,
                trait_sigma=1, comp_sigma = 0.5, dispersal_prob = 0.01, mutation_rate = 1e-6,
                equilib_escape = 1, num_basepairs = 500, alpha = 10000,
                init_type = 'bridge_island', niter = 100000, niterTimestep = 10000)
model <- runRole(roleModel(p))
```

The this `getSumStats(model)` and some genetic diversity and trait Hill's for some timesteps will be 'NaN'. It is non-deterministic, and we assume it's because perhaps there is no variation in these values at a given timestep?

**Comment by isaacovercast** [comment]
Indeed, for the genetic diversity hill #s if there are any '0' in the vector of raw pi values within a given timestep then this fails because of `log(0)`:

```
hill[q == 1] <- exp(sum(-n * log(n)))
```

'q' values other than 1 are perfectly happy with zeros in the pi vector.

**Comment by isaacovercast** [comment]
For trait hills, you get 'NaN' in the case where there is only 1 species present in the local community. If there's only 1 species present then in `.hillDivTrait()` the distance matrix between traits (`dij <- as.matrix(dist(traits))`) is a single value == 0, which eventually leads to a divide-by-zero 2 lines later.

**Comment by diazrenata** [comment]
After wrestling (philosophically) with whether it makes sense to try to hatch a trait diversity for when there is a single species, or whether to simply build in a catch to send single-species communities to evaluate to 0, I'm leaning towards just building in the catch sending single-species to evaluate to zero.

Considerations:
- Functional diversity hill numbers are based on species-distance matrices, which just can't be evaluated for a single species.
- In roleR, we _do_ end up with trait variation in the absence of species variation (intraspecific variation) - but then that's summarized to a species mean anyway, and it seems a little out-of-scope to come up with a new version of trait diversity that looks both intra and inter specifically.

**Comment by diazrenata** [comment]
Amending: I think it should evaluate to 1, yes? Weighting by trait diversity is moot in this instance, so it should converge back to richness.

**Comment by diazrenata** [comment]
See b0ddd3f3339a12726d11bc59c04cdf9bb041e7b8. I feel like there's room for spirited discussion re: the interpretation of trait hill numbers in a single-species community, though! :P

**Comment by isaacovercast** [comment]
agreed. In the case of no trait variation it should be 1.

**Comment by diazrenata** [comment]
Getting sneakier,  b1ba083f7e1e9326420eaccd581322718ce6d6a3. This one catches both edge cases: if there is only one species, or if there are multiple species all with the same trait. (Seems extreme but possible in special circumstances, like a super narrow optimum). In this instance, it effectively reverts to abundance hill numbers.

**Comment by diazrenata** [comment]
(hillR::hill_func just fails in this instance, btw)

**Comment by isaacovercast** [comment]
IDK, I think _conceptuall_ if there is literally zero trait variation then I would say it should be 1. If you think about it from the perspective of # of 'effective species traits' then it would be 1.

**Comment by ajrominger** [comment]
Agreed that for "effective number of species" (i.e. weighted Hill) the answer should be 1, both from the perspective of 1 species, or multiple species but with no trait variation.

I went deep down a rabbit hole on this (nothing against rabbits!)

So the short answer is that *unweighted* hill goes to 0 both for 1 species and multiple species with no trait variance. (Script below showing that).

But weighted hill is $(D_{hill} / Q)^{1/2}$.  And Q in the equation goes to 0 just as fast as $D_{hill}$ (where $D_{hill}$ is unweighted hill).  So again weighted Hill in the limit of 0 trait variation (*either* from 1 species or multiple spp with same trait) goes to 1.

Here's the script

```
# a function to explore limits of trait hill number
# sets q = 2 w/o loss of generality
# n = vector of spp abundances
# trtDiag =  intra-specific trait difference (in reality always 0, but we allow
#            it to be != 0 to explore limit)
# trtOff  =  inter-specific trait differences (i.e. "off diagnal" entries in a
#            distance matrix)

hillLim <- function(n, trtDiag, trtOff) {
    p <- n / sum(n)
    dij <- matrix(trtOff, nrow = length(p), ncol = length(p))
    diag(dij) <- trtDiag
    
    Q <- as.vector(p %*% dij %*% p)
    a <- outer(p, p, '*') / Q
    
    return(sum(dij * a^2)^(1 / (1 - 2)))
}


# explore limit as inter-specific differences go to 0
hillLim(rep(1, 5), trtDiag = 0, trtOff = 1)
## 20

hillLim(rep(1, 5), trtDiag = 0, trtOff = 0.1)
## 2

hillLim(rep(1, 5), trtDiag = 0, trtOff = 0.01)
## 0.2

hillLim(rep(1, 5), trtDiag = 0, trtOff = 0.001)
## 0.02
```

So this says that as trait variation between species goes to 0, so too does Hill diversity go to 0.  **But again, this is for unweighted hill**.

Now for a single species:

```
# explore limit as diagonal element approaches 0 when there is only 1 species
# (i.e. figure out what hill div should be for only 1 species)
hillLim(1, trtDiag = 0.1, trtOff = 1)
## 0.1

hillLim(1, trtDiag = 0.01, trtOff = 1)
## 0.01

hillLim(1, trtDiag = 0.001, trtOff = 1)
## 0.001
```

Here again, for a single species Hill diversity goes to 0.

**But again** these calculations are for **unweighted** Hill.  For weighted Hill, I convinced myself the answer is indeed equal to 1 because D_hill and Q go to 0 at the same rate.

**Comment by ajrominger** [comment]
*OH!* Something important though that's coming out of this is: right now we're returning unweighted Hill numbers right?  So is that what we want to do, or do we want to switch to returning weighted Hills?

**Comment by ajrominger** [comment]
Damn! I been working on this more and I was **wrong** about weighted trait Hill for the case no variation: weighted trait Hill for the case with no variation might equal weighted abundance-based Hill with the same SAD...still working on it, stay tuned, or if you all have it figured out, even better!

**Comment by ajrominger** [comment]
One thing for sure though we have to pick a column (attribute diversity or generalized hill):

<img width="695" alt="image" src="https://github.com/role-model/roleR/assets/2481785/0791c9a8-9045-43a5-a48e-ee408441f2cf">

**Comment by isaacovercast** [comment]
Generalized Hills, that's my vote. The interpretation is very natural.

**Comment by diazrenata** [comment]
I agree with Isaac about generalized hills.

**Comment by ajrominger** [comment]
also agree with both of you about using generalized hills!

so this means we need to revise the trait and phylo hill numbers to return the generalized value (currently they return the "attribute diversity" from Chao's table). the abundance hill function (and thus the gen div hill) returns the generalized value because the generalized and attribute div for abundance are equivalent

**Comment by ajrominger** [comment]
and my final thought on hill numbers when trait variation is absolutely 0 is that the generalized number should indeed be 1.  if you all are good to go with that, let's go for it

**Comment by diazrenata** [comment]
Unless I'm mistaken, the trait hill number function is already returning generalized hills! e4b7a4701fa52e80697fa8a8a750de8fd9cbfa99 implements a check if the trait diversity is 0 (which would occur if species richness is 0, or if trait diversity is 0) and returns 1 for all q. 

I'll work on phylo next!

**Comment by ajrominger** [comment]
right you are about generalized hills! don't know how i convinced myself otherwise!
